notebook/terminal/terminalmanager.py

```diff
@@ -0,0 +1,150 @@
+
+
+from datetime import timedelta
+from notebook._tz import utcnow, isoformat
+from terminado import NamedTermManager
+from tornado import web
+from tornado.ioloop import IOLoop, PeriodicCallback
+from traitlets import Integer
+from traitlets.config import Configurable
+from ..prometheus.metrics import TERMINAL_CURRENTLY_RUNNING_TOTAL
+
+
+class TerminalManager(Configurable, NamedTermManager):
+    """  """
+
+    _culler_callback = None
+
+    _initialized_culler = False
+
+    cull_inactive_timeout = Integer(0, config=True,
+                                    help="""Timeout (in seconds) in which a terminal has been inactive and ready to be culled.
+        Values of 0 or lower disable culling."""
+                                    )
+
+    cull_interval_default = 300  # 5 minutes
+    cull_interval = Integer(cull_interval_default, config=True,
+        help="""The interval (in seconds) on which to check for terminals exceeding the inactive timeout value."""
+                            )
+
+    # -------------------------------------------------------------------------
+    # Methods for managing terminals
+    # -------------------------------------------------------------------------
+    def __init__(self, *args, **kwargs):
+        super(TerminalManager, self).__init__(*args, **kwargs)
+
+    def create(self):
+        """Create a new terminal."""
+        name, term = self.new_named_terminal()
+        # Monkey-patch last-activity, similar to kernels.  Should we need
+        # more functionality per terminal, we can look into possible sub-
+        # classing or containment then.
+        term.last_activity = utcnow()
+        model = self.terminal_model(name)
+        # Increase the metric by one because a new terminal was created
+        TERMINAL_CURRENTLY_RUNNING_TOTAL.inc()
+        # Ensure culler is initialized
+        self._initialize_culler()
+        return model
+
+    def get(self, name):
+        """Get terminal 'name'."""
+        model = self.terminal_model(name)
+        return model
+
+    def list(self):
+        """Get a list of all running terminals."""
+        models = [self.terminal_model(name) for name in self.terminals]
+
+        # Update the metric below to the length of the list 'terms'
+        TERMINAL_CURRENTLY_RUNNING_TOTAL.set(
+            len(models)
+        )
+        return models
+
+    async def terminate(self, name, force=False):
+        """Terminate terminal 'name'."""
+        self._check_terminal(name)
+        await super(TerminalManager, self).terminate(name, force=force)
+
+        # Decrease the metric below by one
+        # because a terminal has been shutdown
+        TERMINAL_CURRENTLY_RUNNING_TOTAL.dec()
+
+    async def terminate_all(self):
+        """Terminate all terminals."""
+        terms = [name for name in self.terminals]
+        for term in terms:
+            await self.terminate(term, force=True)
+
+    def terminal_model(self, name):
```

Comment:
*nitpick*, I think we should name this `get_terminal_model(...)`; otherwise this looks like a property and not a method.